accord-core/src/main/java/accord/impl/InMemoryCommandStore.java

```diff
@@ -0,0 +1,371 @@
+    }
+
+    @Override
+    public Command command(TxnId txnId)
+    {
+        return commands.computeIfAbsent(txnId, id -> new InMemoryCommand(this, id));
+    }
+
+    public boolean hasCommand(TxnId txnId)
+    {
+        return commands.containsKey(txnId);
+    }
+
+    @Override
+    public CommandsForKey commandsForKey(Key key)
+    {
+        return commandsForKey.computeIfAbsent(key, InMemoryCommandsForKey::new);
+    }
+
+    public boolean hasCommandsForKey(Key key)
+    {
+        return commandsForKey.containsKey(key);
+    }
+
+    @Override
+    public CommandsForKey maybeCommandsForKey(Key key)
+    {
+        return commandsForKey.get(key);
+    }
+
+    // TODO: command store api will need to support something like this for repair/streaming
+    public void forEpochCommands(KeyRanges ranges, long epoch, Consumer<Command> consumer)
+    {
+        Timestamp minTimestamp = new Timestamp(epoch, Long.MIN_VALUE, Integer.MIN_VALUE, Node.Id.NONE);
+        Timestamp maxTimestamp = new Timestamp(epoch, Long.MAX_VALUE, Integer.MAX_VALUE, Node.Id.MAX);
+        for (KeyRange range : ranges)
+        {
+            Iterable<CommandsForKey> rangeCommands = commandsForKey.subMap(range.start(),
+                                                                           range.startInclusive(),
+                                                                           range.end(),
+                                                                           range.endInclusive()).values();
+            for (CommandsForKey commands : rangeCommands)
+            {
+                commands.forWitnessed(minTimestamp, maxTimestamp, consumer);
+            }
+        }
+    }
+
+    // TODO: command store api will need to support something like this for repair/streaming
+    public void forCommittedInEpoch(KeyRanges ranges, long epoch, Consumer<Command> consumer)
+    {
+        Timestamp minTimestamp = new Timestamp(epoch, Long.MIN_VALUE, Integer.MIN_VALUE, Node.Id.NONE);
+        Timestamp maxTimestamp = new Timestamp(epoch, Long.MAX_VALUE, Integer.MAX_VALUE, Node.Id.MAX);
+        for (KeyRange range : ranges)
+        {
+            Iterable<CommandsForKey> rangeCommands = commandsForKey.subMap(range.start(),
+                                                                           range.startInclusive(),
+                                                                           range.end(),
+                                                                           range.endInclusive()).values();
+            for (CommandsForKey commands : rangeCommands)
+            {
+
+                Collection<Command> committed = commands.committedByExecuteAt()
+                        .between(minTimestamp, maxTimestamp).collect(Collectors.toList());
+                committed.forEach(consumer);
+            }
+        }
+    }
+
+    protected void processInternal(Consumer<? super CommandStore> consumer, Promise<Void> promise)
```

Comment:
nit: "protected void processInternal(Consumer<? super CommandStore> consumer, Promise<Void> promise)" could be the following to avoid copy/paste

```
protected void processInternal(Consumer<? super CommandStore> consumer, Promise<Void> promise)
    {
        processInternal(cs -> {
            consumer.accept(cs);
            return null;
        }, promise);
    }
```

Reply:
resolved